File: example/pages/TodoPage.ts
import { BasePage } from 'conductor-e2e';
import type { Locator } from 'playwright';

export class TodoPage extends BasePage {
  private readonly titleInput: Locator;
  private readonly addButton: Locator;
  private readonly todoList: Locator;
  private readonly prioritySelect: Locator;

  constructor(...args: ConstructorParameters<typeof BasePage>) {
    super(...args);
    this.titleInput = this.page.locator('[data-testid="todo-input"]');
    this.addButton = this.page.locator('[data-testid="todo-add"]');
    this.todoList = this.page.locator('[data-testid="todo-list"]');
    this.prioritySelect = this.page.locator('[data-testid="todo-priority-select"]');
  }

  async createTodo(title: string, priority?: string): Promise<void> {
    await this.titleInput.fill(title);
    if (priority) {
      await this.prioritySelect.selectOption(priority);
    }
    await this.addButton.click();
    await this.page.waitForLoadState('networkidle');
  }

  async assertVisible(title: string): Promise<void> {
    await this.todoList.locator(`text=${title}`).waitFor({ state: 'visible' });
  }

  async assertNotVisible(title: string): Promise<void> {
    await this.todoList.locator(`text=${title}`).waitFor({ state: 'hidden', timeout: 5000 }).catch(() => {});
    const count = await this.todoList.locator(`[data-testid="todo-item"]`).filter({ hasText: title }).count();
    if (count > 0) throw new Error(`Todo "${title}" is still visible but should not be`);
  }

  async getTodoCount(): Promise<number> {
    return this.todoList.locator('[data-testid="todo-item"]').count();
  }

  private getTodoItem(title: string): Locator {
    return this.todoList.locator('[data-testid="todo-item"]').filter({ hasText: title });
  }

  async editTodo(currentTitle: string, newTitle: string): Promise<void> {
    this.page.once('dialog', async (dialog) => {
      await dialog.accept(newTitle);
    });
    await Promise.all([
      this.page.waitForNavigation({ waitUntil: 'networkidle' }),
      this.getTodoItem(currentTitle).locator('[data-testid="todo-edit"]').click(),
    ]);
  }

  async deleteTodo(title: string): Promise<void> {
    this.page.once('dialog', async (dialog) => {
      await dialog.accept();
    });
    await Promise.all([
      this.page.waitForNavigation({ waitUntil: 'networkidle' }),
      this.getTodoItem(title).locator('[data-testid="todo-delete"]').click(),
    ]);
  }

  async toggleTodo(title: string): Promise<void> {
    await Promise.all([
      this.page.waitForNavigation({ waitUntil: 'networkidle' }),
      this.getTodoItem(title).locator('[data-testid="todo-toggle"]').click(),
    ]);
  }

  async assertStatus(title: string, status: string): Promise<void> {
    const item = this.getTodoItem(title);
    await item.locator(`.todo-status`).waitFor({ state: 'visible' });
    const text = await item.locator(`.todo-status`).textContent();
    if (!text?.includes(status)) {
      throw new Error(`Expected todo "${title}" to have status "${status}" but got "${text}"`);
    }
  }

  async assertPriority(title: string, priority: string): Promise<void> {
    const item = this.getTodoItem(title);
    const text = await item.locator('[data-testid="todo-priority"]').textContent();
    if (text?.trim() !== priority) {
      throw new Error(`Expected todo "${title}" to have priority "${priority}" but got "${text}"`);
    }
  }

}

File: java/conductor-example/src/test/java/com/nouhouari/conductor/example/pages/TodoPage.java
package com.nouhouari.conductor.example.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.TimeoutError;
import com.microsoft.playwright.options.LoadState;
import com.microsoft.playwright.options.WaitForSelectorState;
import com.microsoft.playwright.options.WaitUntilState;
import com.nouhouari.conductor.config.EnvironmentConfig;
import com.nouhouari.conductor.pages.BasePage;

/** Java port of example/pages/TodoPage.ts. */
public class TodoPage extends BasePage {

    private final Locator titleInput;
    private final Locator addButton;
    private final Locator todoList;
    private final Locator prioritySelect;

    public TodoPage(Page page, EnvironmentConfig config) {
        super(page, config);
        this.titleInput = page.locator("[data-testid=\"todo-input\"]");
        this.addButton = page.locator("[data-testid=\"todo-add\"]");
        this.todoList = page.locator("[data-testid=\"todo-list\"]");
        this.prioritySelect = page.locator("[data-testid=\"todo-priority-select\"]");
    }

    public void createTodo(String title) {
        createTodo(title, null);
    }

    public void createTodo(String title, String priority) {
        titleInput.fill(title);
        if (priority != null) {
            prioritySelect.selectOption(priority);
        }
        addButton.click();
        page.waitForLoadState(LoadState.NETWORKIDLE);
    }

    public void assertVisible(String title) {
        todoList.locator("text=" + title).waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE));
    }

    public void assertNotVisible(String title) {
        try {
            todoList.locator("text=" + title).waitFor(new Locator.WaitForOptions()
                    .setState(WaitForSelectorState.HIDDEN).setTimeout(5000));
        } catch (TimeoutError ignored) {
            // mirrors the TS side's .catch(() => {}) — fall through to the count check below
        }
        int count = todoList.locator("[data-testid=\"todo-item\"]").filter(new Locator.FilterOptions().setHasText(title)).count();
        if (count > 0) {
            throw new AssertionError("Todo \"" + title + "\" is still visible but should not be");
        }
    }

    public int getTodoCount() {
        return todoList.locator("[data-testid=\"todo-item\"]").count();
    }

    private Locator getTodoItem(String title) {
        return todoList.locator("[data-testid=\"todo-item\"]").filter(new Locator.FilterOptions().setHasText(title));
    }

    public void editTodo(String currentTitle, String newTitle) {
        page.onceDialog(dialog -> dialog.accept(newTitle));
        page.waitForNavigation(new Page.WaitForNavigationOptions().setWaitUntil(WaitUntilState.NETWORKIDLE),
                () -> getTodoItem(currentTitle).locator("[data-testid=\"todo-edit\"]").click());
    }

    public void deleteTodo(String title) {
        page.onceDialog(com.microsoft.playwright.Dialog::accept);
        page.waitForNavigation(new Page.WaitForNavigationOptions().setWaitUntil(WaitUntilState.NETWORKIDLE),
                () -> getTodoItem(title).locator("[data-testid=\"todo-delete\"]").click());
    }

    public void toggleTodo(String title) {
        page.waitForNavigation(new Page.WaitForNavigationOptions().setWaitUntil(WaitUntilState.NETWORKIDLE),
                () -> getTodoItem(title).locator("[data-testid=\"todo-toggle\"]").click());
    }

    public void assertStatus(String title, String status) {
        Locator item = getTodoItem(title);
        item.locator(".todo-status").waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE));
        String text = item.locator(".todo-status").textContent();
        if (text == null || !text.contains(status)) {
            throw new AssertionError("Expected todo \"" + title + "\" to have status \"" + status + "\" but got \"" + text + "\"");
        }
    }

    public void assertPriority(String title, String priority) {
        Locator item = getTodoItem(title);
        String text = item.locator("[data-testid=\"todo-priority\"]").textContent();
        if (text == null || !text.trim().equals(priority)) {
            throw new AssertionError("Expected todo \"" + title + "\" to have priority \"" + priority + "\" but got \"" + text + "\"");
        }
    }
}
